Configure now playing media session to support external playback

rsfuller opened the pull request:
#### d1ed61f568c7dbcfca9b7f7f535618d03b6fa820
<pre>
Configure now playing media session to support external playback
<a href="https://bugs.webkit.org/show_bug.cgi?id=287684">https://bugs.webkit.org/show_bug.cgi?id=287684</a>
<a href="https://rdar.apple.com/137156850">rdar://137156850</a>

Reviewed by Andy Estes.

Add a new visionOS WKWebView method to enable configuring the now playing media
session for external playback. This puts the visionOS media player into a
custom presentation (external) and returns the media player&apos;s view controller
to the caller who has the responsibility to present it. It also provides a method
to exit this external playback.

On entering presentation, it sets the playerIdentifier for the video element
which normally gets set as part of the fullscreen flow but we are bypassing that
route in this mode -- without this the media player will not be able to render.
It also puts LinearMediaPlayer in a new presentationState (.external) which
allows WebKit to configure the player as needed without affecting video playing
outside of this feature. Finally, it makes and returns the player view
controller.

On exiting presentation, it sets the state of linearMediaPlayer back to .inline
presentation and tears down the configuration. Importantly, it clears the video
receiver endpoint to allow the media session to render in its non-LMK
presentation.

While this mode is similar conceptually to fullscreen, it intentionally does not
latch onto the fullscreen element state, as it should be a separate mode and the
media element should not falsely believe it is in a fullscreen presentation.
While in external presentation, fullscreen mode is prevented from being entered.

Note: Remove LinearMediaPlayer&apos;s fullscreenSceneBehavior property until it is
able to be published (<a href="https://rdar.apple.com/122435030">rdar://122435030</a>).

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::setPlayerIdentifierForVideoElement):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::setPlayerIdentifierForVideoElement):
* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::setPlayerIdentifierForVideoElement):
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.mm:
(WebCore::PlaybackSessionInterfaceIOS::setVideoPresentationInterface):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
(WebCore::VideoPresentationInterfaceIOS::VideoPresentationInterfaceIOS):
(WebCore::VideoPresentationInterfaceIOS::enterExternalPlayback):
(WebCore::VideoPresentationInterfaceIOS::exitExternalPlayback):
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
(-[WKLinearMediaPlayerDelegate linearMediaPlayerClearVideoReceiverEndpoint:]):
(WebKit::PlaybackSessionInterfaceLMK::supportsLinearMediaPlayerChanged):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.h:
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::enterExternalPlayback):
(WebKit::VideoPresentationInterfaceLMK::exitExternalPlayback):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _enterExternalPlaybackForNowPlayingMediaSessionWithCompletionHandler:]):
(-[WKWebView _exitExternalPlaybackWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::setPlayerIdentifierForVideoElement):
(WebKit::PlaybackSessionManagerProxy::setPlayerIdentifierForVideoElement):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::enterExternalPlaybackForNowPlayingMediaSession):
(WebKit::WebPageProxy::exitExternalPlayback):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setPlayerIdentifierForVideoElement):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(SwiftOnlyData.fullscreenSceneBehaviors):
(WKSLinearMediaPlayer.enterExternalPresentation(_:(any Error)?) -&gt; Void:)):
(WKSLinearMediaPlayer.exitExternalPresentation(_:(any Error)?) -&gt; Void:)):
(WKSLinearMediaPlayer.enterFullscreen(_:(any Error)?) -&gt; Void:)):
(WKSLinearMediaPlayer.exitFullscreen(_:(any Error)?) -&gt; Void:)):
(WKSLinearMediaPlayer.presentationStateChanged(_:)):
(WKSLinearMediaPlayer.fullscreenSceneBehaviorsPublisher):
(WKSLinearMediaPlayer.toggleInlineMode):
(WKSLinearMediaPlayer.willEnterFullscreen):
(WKSLinearMediaPlayer.willExitFullscreen):
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
(WKSLinearMediaPresentationState.description):
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaPlayer.h:
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::setPlayerIdentifierForVideoElement):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.messages.in:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionManager::setPlayerIdentifierForVideoElement):
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:
(WebKit::VideoPresentationManager::setPlayerIdentifierForVideoElement):

Canonical link: <a href="https://commits.webkit.org/290837@main">https://commits.webkit.org/290837@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/74bee0590eb710ba352141729d2d4570e7d8c05c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91232 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10771 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/270 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96228 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/41972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93282 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11167 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19089 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70096 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/41972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94233 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/8511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/82651 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50422 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/8283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/228 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41110 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/78603 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/241 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98215 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18421 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79103 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18677 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78483 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78306 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19364 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22816 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11595 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18421 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/23725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18142 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/21602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/19915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -7295,6 +7295,24 @@ bool HTMLMediaElement::videoUsesElementFullscreen() const
+    if (!window)
+        return;
+
+    if (auto* asVideo = dynamicDowncast<HTMLVideoElement>(*this)) {
```
```suggestion
    if (RefPtr asVideo = dynamicDowncast<HTMLVideoElement>(*this)) {
```

aestes commented on Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
```diff
@@ -175,6 +175,7 @@ void togglePictureInPicture() override { }
     void enterInWindowFullscreen() override { }
     void exitInWindowFullscreen() override { }
     void enterFullscreen() override { }
+    void setPlayerIdentifierForVideoElement() override { }
```
Nit: I think this can be `final` (although I know we mistakenly use `override` elsewhere).

aestes commented on Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
```diff
@@ -358,6 +365,12 @@ - (void)linearMediaPlayer:(WKSLinearMediaPlayer *)player setVideoReceiverEndpoin
+        // If the player is in external presentation (which uses LinearMediaPlayer) but the current
+        // media engine does not support it, exit external presentation.
+        if (RefPtr strongVideoPresentationInterface = m_videoPresentationInterface.get())
+            strongVideoPresentationInterface->exitExternalPlayback([](bool) { });
```
```suggestion
        if (RefPtr videoPresentationInterface = m_videoPresentationInterface.get())
            videoPresentationInterface->exitExternalPlayback([](bool) { });
```

aestes commented on Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
```diff
@@ -525,7 +563,7 @@ extension WKSLinearMediaPlayer: @retroactive Playable {
     public var fullscreenSceneBehaviorsPublisher: AnyPublisher<[FullscreenBehaviors], Never> {
         // FIXME: Publish fullscreenSceneBehaviors once rdar://122435030 is resolved
-        Just(FullscreenBehaviors.default).eraseToAnyPublisher()
+        swiftOnlyData.$fullscreenSceneBehaviors.eraseToAnyPublisher()
```
This doesn't seem quite right, because there is already a `fullscreenSceneBehaviors` property declared in WKSLinearMediaPlayer.h. Instead off adding a new property to `SwiftOnlyData`, why don't we use the existing `fullscreenSceneBehaviors` property and add a convenience property that maps from `WKSLinearMediaFullscreenBehaviors` to `FullscreenBehaviors` (see `contentTypePublisher` for an example).

rsfuller replied:
We are limited by rdar://122435030, in that LMK takes an array of OptionSets and does not properly check the stream of values. So for now, move to using a `CurrentValueSubject` that is functionally equivalent to the existing `Just` in that it publishes a single value, but allows us to update it with `send()` when we need to change it.

As well, we decided to remove `fullscreenSceneBehaviors` property from LinearMediaPlayer so it doesn't cause confusion / accidental adoption while we are not currently publishing it.

aestes reviewed (approved):
Looks great! r=me with some comments.

rsfuller committed:
Configure now playing media session to support external playback
https://bugs.webkit.org/show_bug.cgi?id=287684
rdar://137156850

Reviewed by Andy Estes.

Add a new visionOS WKWebView method to enable configuring the now playing media
session for external playback. This puts the visionOS media player into a
custom presentation (external) and returns the media player's view controller
to the caller who has the responsibility to present it. It also provides a method
to exit this external playback.

On entering presentation, it sets the playerIdentifier for the video element
which normally gets set as part of the fullscreen flow but we are bypassing that
route in this mode -- without this the media player will not be able to render.
It also puts LinearMediaPlayer in a new presentationState (.external) which
allows WebKit to configure the player as needed without affecting video playing
outside of this feature. Finally, it makes and returns the player view
controller.

On exiting presentation, it sets the state of linearMediaPlayer back to .inline
presentation and tears down the configuration. Importantly, it clears the video
receiver endpoint to allow the media session to render in its non-LMK
presentation.

While this mode is similar conceptually to fullscreen, it intentionally does not
latch onto the fullscreen element state, as it should be a separate mode and the
media element should not falsely believe it is in a fullscreen presentation.
While in external presentation, fullscreen mode is prevented from being entered.

Note: Remove LinearMediaPlayer's fullscreenSceneBehavior property until it is
able to be published (rdar://122435030).

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::setPlayerIdentifierForVideoElement):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::setPlayerIdentifierForVideoElement):
* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::setPlayerIdentifierForVideoElement):
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.mm:
(WebCore::PlaybackSessionInterfaceIOS::setVideoPresentationInterface):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
(WebCore::VideoPresentationInterfaceIOS::VideoPresentationInterfaceIOS):
(WebCore::VideoPresentationInterfaceIOS::enterExternalPlayback):
(WebCore::VideoPresentationInterfaceIOS::exitExternalPlayback):
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
(-[WKLinearMediaPlayerDelegate linearMediaPlayerClearVideoReceiverEndpoint:]):
(WebKit::PlaybackSessionInterfaceLMK::supportsLinearMediaPlayerChanged):
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.h:
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::enterExternalPlayback):
(WebKit::VideoPresentationInterfaceLMK::exitExternalPlayback):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _enterExternalPlaybackForNowPlayingMediaSessionWithCompletionHandler:]):
(-[WKWebView _exitExternalPlaybackWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::setPlayerIdentifierForVideoElement):
(WebKit::PlaybackSessionManagerProxy::setPlayerIdentifierForVideoElement):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::enterExternalPlaybackForNowPlayingMediaSession):
(WebKit::WebPageProxy::exitExternalPlayback):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setPlayerIdentifierForVideoElement):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(SwiftOnlyData.fullscreenSceneBehaviors):
(WKSLinearMediaPlayer.enterExternalPresentation(_:(any Error)?) -> Void:)):
(WKSLinearMediaPlayer.exitExternalPresentation(_:(any Error)?) -> Void:)):
(WKSLinearMediaPlayer.enterFullscreen(_:(any Error)?) -> Void:)):
(WKSLinearMediaPlayer.exitFullscreen(_:(any Error)?) -> Void:)):
(WKSLinearMediaPlayer.presentationStateChanged(_:)):
(WKSLinearMediaPlayer.fullscreenSceneBehaviorsPublisher):
(WKSLinearMediaPlayer.toggleInlineMode):
(WKSLinearMediaPlayer.willEnterFullscreen):
(WKSLinearMediaPlayer.willExitFullscreen):
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
(WKSLinearMediaPresentationState.description):
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaPlayer.h:
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::setPlayerIdentifierForVideoElement):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.messages.in:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionManager::setPlayerIdentifierForVideoElement):
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.mm:
(WebKit::VideoPresentationManager::setPlayerIdentifierForVideoElement):

Canonical link: https://commits.webkit.org/290837@main
